code/controllers/subsystem/time_track.dm

```diff
@@ -24,12 +32,23 @@ SUBSYSTEM_DEF(time_track)
 
 	if (!first_run)
 		var/tick_drift = max(0, (((current_realtime - last_tick_realtime) - (current_byondtime - last_tick_byond_time)) / world.tick_lag))
-
 		time_dilation_current = tick_drift / (current_tickcount - last_tick_tickcount) * 100
 
 		time_dilation_avg_fast = MC_AVERAGE_FAST(time_dilation_avg_fast, time_dilation_current)
 		time_dilation_avg = MC_AVERAGE(time_dilation_avg, time_dilation_avg_fast)
 		time_dilation_avg_slow = MC_AVERAGE_SLOW(time_dilation_avg_slow, time_dilation_avg)
+		// Time Dilation tick_lag adjustment code, credit to MyBlueCorners and the rest of the goonstation dev team!
+		var/last_interval_tick_offset = max(0, (current_realtime - last_tick_realtime) - (current_byondtime - last_tick_byond_time))
+
+		var/dilated_tick_lag = world.tick_lag
+		if(last_interval_tick_offset >= TICKLAG_DILATION_THRESHOLD)
+			dilated_tick_lag = min(world.tick_lag + TICKLAG_DILATION_INC, OVERLOADED_WORLD_TICKLAG)
+		else if(last_interval_tick_offset <= TICKLAG_NORMALIZATION_THRESHOLD)
+			dilated_tick_lag = max(world.tick_lag - TICKLAG_DILATION_DEC, MIN_TICKLAG)
+
+		if(world.tick_lag != dilated_tick_lag)
+			world.change_tick_lag(dilated_tick_lag)
```

Comment:
This is being quite taxing on the timer SS.
You should avoid changing this value unless with very good reasons.